src/Analyzer.JsonRuleEngine/Converters/ExpressionConverter.cs

```diff
@@ -70,6 +59,10 @@ public override ExpressionDefinition ReadJson(JsonReader reader, Type objectType
                 {
                     return CreateExpressionDefinition<AllOfExpressionDefinition>(jsonObject, serializer);
                 }
+                else if (objectPropertyNames.Contains("anyOf", StringComparer.OrdinalIgnoreCase))
```

Comment:
I suggest doing the structured expression validations here instead of in `ValidateExpressions`.  Instead of returning the created object directly, create the object, validate it, then return it.  You should also remove the surrounding `if` statement, since you have to check for each expression again anyway.

Reply:
Is this what you're recommending?
``` csharp
public override ExpressionDefinition ReadJson(JsonReader reader, Type objectType, [AllowNull] ExpressionDefinition existingValue, bool hasExistingValue, JsonSerializer serializer)
{
    if (reader.TokenType == JsonToken.Null)
    {
        return null;
    }

    var jsonObject = JObject.Load(reader);

    var objectPropertyNames = jsonObject.Properties().Select(property => property.Name).ToList();

    var expressionJsonPropertyNames = GetExpressionJsonPropertyNames();
    var structuredExpressions = GetStructuredExpressionJsonPropertyNames();

    var expressionPropertyCount = objectPropertyNames.Count(property => expressionJsonPropertyNames.Contains(property));
    if (expressionPropertyCount == 1)
    {
        // See if a property representing a structured expression is present.  If so, parse that structured expression.
        var structuredExpressionPropertyName = objectPropertyNames.FirstOrDefault(property => structuredExpressions.Contains(property));
        if (!string.IsNullOrEmpty(structuredExpressionPropertyName))
        {
            var structuredExpression = jsonObject.InsensitiveToken(structuredExpressionPropertyName);

            // If there are no leaf expressions defined in 
            // the structured expression, throw an error
            if (structuredExpression.Count() == 0)
            {
                throw new JsonException($"No leaf expressions were specified in the {structuredExpression} expression");
            }

            if (objectPropertyNames.Contains("allOf", StringComparer.OrdinalIgnoreCase))
            {
                return CreateExpressionDefinition<AllOfExpressionDefinition>(jsonObject, serializer);
            }
            else if (objectPropertyNames.Contains("anyOf", StringComparer.OrdinalIgnoreCase))
            {
                return CreateExpressionDefinition<AnyOfExpressionDefinition>(jsonObject, serializer);
            }

            throw new JsonException($"Expression is not supported. One of the following fields is not supported: {string.Join(", ", objectPropertyNames.ToArray())}");
        }
        else
        {
            return CreateExpressionDefinition<LeafExpressionDefinition>(jsonObject, serializer);
        }
    }

    throw new JsonException(expressionPropertyCount > 1 ?
        $"Too many expressions specified in evaluation.  Only one is allowed.  Original JSON: {jsonObject}" :
        $"Invalid evaluation in JSON.  No expressions are specified (must specify exactly one).  Original JSON: {jsonObject}");
}
```

Reply:
Closer to this:
``` c#
var expressionPropertyCount = ...;
if (expressionPropertyCount == 1)
{
    // See if a property representing a structured expression is present.  If so, parse that structured expression.
    if (objectPropertyNames.Contains("allOf", StringComparer.OrdinalIgnoreCase))
    {
        var allOfExpression = CreateExpressionDefinition<AllOfExpressionDefinition>(jsonObject, serializer);
        if (!(allOfExpression.allOf?.Length > 0))
        {
            // exception
        }
        if (allOfExpression.allOf.Count(e => e == null) > 0)
        {
            // exception
        }
        return allOfExpression;
    }
    // again for anyOf
}
// else create leaf

// exception otherwise
```
It might even be good to put the validation of the leaf arrays in a function within the respective classes, which would return itself or throw an exception, so parsing could look something like...
``` c#
if (objectPropertyNames.Contains("allOf", StringComparer.OrdinalIgnoreCase))
{
    return CreateExpressionDefinition<AllOfExpressionDefinition>(jsonObject, serializer).Validate();
}
```